### PR TITLE
Escape textual ion symbols less often

### PR DESCRIPTION
**Pull request by joshterrell805**
I didn't find a definitive list of keywords. I'm not sure if there are any others besides `null, nan, false, true`. I'd expect `inf` to be a keyword, but it worked out okay without quoting.

Found this: https://github.com/amzn/ion-python/blob/master/amazon/ion/reader_text.py#L1279-L1289

### CODE REVIEW

File: tests/writer_util.py [hunk 1]
```diff
@@ -60,9 +60,14 @@ def _scalar_p(ion_type, value, expected, force_stream_end):
...
 
 
-def _convert_symbol_pairs(symbol_pairs):
+def _convert_symbol_pairs_to_string_pairs(symbol_pairs):
     for value, literal in symbol_pairs:
-        yield (value, literal.replace(b"'", b'"'))
+        if literal.decode('utf-8')[0] == "'":
+            yield (value, literal.replace(b"'", b'"'))
+        else:
+            # there's one symbol in our symbol table which isn't quoted
```
[the remark this answers is not in the document]
Comment (joshterrell805) [reply tests/writer_util.py]: That makes sense, silly me :)

File: tests/test_writer_text.py [hunk 2]
```diff
@@ -237,3 +239,53 @@ def test_valid_indent_strs(p):
...
+        _P_Q('a.0', True),
+        _P_Q('+inf', True),
+        _P_Q('-inf', True),
+        _P_Q('null.int', True),
+        _P_Q('1', True),
+        _P_Q('{', True),
+        _P_Q('\'', True, backslash_required=True),
+        _P_Q('"', True),
+        _P_Q('\\', True, backslash_required=True),
+        # begin reserved words
```
[the remark this answers is not in the document]
Comment (joshterrell805) [reply tests/test_writer_text.py]: Alright, I'll add them as well

File: tests/test_simpleion.py [hunk 3]
```diff
@@ -507,16 +508,21 @@ class PrettyPrintParams(record('ion_text', 'indent', ('exact_text', None), ('nee
...
         PrettyPrintParams(ion_text='[a, b, chair::2008-08-08T]', indent=None, # not pretty print
-            exact_text="$ion_1_0 ['a','b','chair'::2008-08-08T]"),
+            exact_text="$ion_1_0 [a,b,chair::2008-08-08T]"),
         PrettyPrintParams(ion_text='[apple, {roof: false}]', indent='\t',
-            exact_text="$ion_1_0\n[\n\t'apple',\n\t{\n\t\t'roof': false\n\t}\n]"),
+            exact_text="$ion_1_0\n[\n\tapple,\n\t{\n\t\troof: false\n\t}\n]"),
+        PrettyPrintParams(ion_text='[apple, "banana", {roof: false}]', indent='\t',
+            exact_text="$ion_1_0\n[\n\tapple,\n\t\"banana\",\n\t{\n\t\troof: false\n\t}\n]"),
         PrettyPrintParams(ion_text='[apple, {roof: false, walls:4, door: wood::large::true}]', indent='\t',
-            needles=["$ion_1_0\n[\n\t'apple',\n\t{", "\n\t}\n]"],
```
[the remark this answers is not in the document]
Comment (joshterrell805) [reply tests/test_simpleion.py]: Oh weird, looks like I accidentally removed this. Yeah if it's just being used for one test case, I'll remove the needles.

File: amazon/ion/writer_text.py [hunk 4]
```diff
@@ -241,19 +241,23 @@ def _bytes_text(code_point_iter, quote, prefix=b'', suffix=b''):
 
 _SINGLE_QUOTE = b"'"
 _DOUBLE_QUOTE = b'"'
+_UNQUOTED_SYMBOL_REGEX = re.compile(r'\A[a-zA-Z$_][a-zA-Z0-9$_]*\Z')
+_SYMBOLS_REQUIRING_QUOTES = ['nan', 'null', 'false', 'true']
```
[the remark this answers is not in the document]
Comment (joshterrell805) [reply amazon/ion/writer_text.py]: Hmm yes faster, more to maintain, but I think clearer perhaps. I'll take another pass at this. Thanks for the feedback!

**Comment by raganhan** [comment]
@joshterrell805 can you confirm if you made the changes that were asked by @tgregg? It seems like you did since the some comments are marked as outdated and all were marked as resolved, but there is only one commit in the PR history. 

Just want to double check before merging the change

**Comment by joshterrell805** [comment]
Hi @raganhan, I addressed all comments, but I pushed a forced revision which is why there's only one commit. For the following comment:

> The other reserved words are +inf, -inf, and all of the typed nulls, e.g. null.int. However, all of these are obviously flagged by _UNQUOTED_SYMBOL_REGEX, so the code is still correct as-is. Do you think it's faster to look them up from _SYMBOLS_REQUIRING_QUOTES?

Instead of adding those symbols to the `_SYMBOLS_REQUIRING_QUOTES`, list, I renamed that list and added a comment making the semantics clearer.

**Comment by raganhan** [comment]
@joshterrell805 sounds good, merging 

Thanks for the contribution :)

**Pull request merged**